### ISSUE TITLE
Regression tests are running slower than previously

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#622]
Regression tests are now running by almost a facto of 2 slower since we have rearranged the atomic data files.  We need to understand whether this is due to the data files themselves, or to some other change that occurred about the same time (most likely some changes associated with how we are doing integrations).  

The problem appears to have developed between these two builds

**Has problem**

commit 66ca0cc3fc20c385965d2574d32bb9a3443d8cc3
Author: higginbottom <nick_higginbottom@fastmail.fm>
Date:   Fri Oct 11 12:21:46 2019 +0100

    Modified examples files for new data files

**Does not have problem**

commit 77eac957497f70e0c5ae668c6a3fa987fd3d1428
Merge: f5d682d9 ebb11c8c
Author: Knox S. Long <long@stsci.edu>
Date:   Mon Oct 7 12:05:17 2019 -0500

    Merge branch 'dev' of https://github.com/agnwinds/python into dev

**Comment by kslong** [comment]
OK, I have verified two items

- There are no obvious changes between the two commits above except for the changes associated with renaming the atomic data files in the .pf files
- If one runs the current version of the code with the old datafiles, the run times are back to what they were before the old data files were removed.

So this implies that all of the speed losses were directly due to the differences in the atomic data files, as @Higginbottom suspected.  

The only remaining questions are 

- What is so different about standard78 from standard80 that affects the run time so much?  Nick, some indication of what is different here would be helpful
- Given this change, are there portions of the code that we need to look at from the point of view of speeding up?
- Should we be leaving some of the earlier files in the repostiory?

**Comment by Higginbottom** [comment]
|   |   |   |   |   | qromb old data | gsl romb old data | gsl romb new data | gsl qags new data | gsl_qags with BU new data | gsl_qags with BU old data
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
cv_qromb |   |   |   |   |   | 0 | 0 | 0 | 0 | 0 | 0
Initializing | variables | for | cv_qromb |   |   | 0 | 0 | 0 | 0 | 0 | 0
Finished | initialization | for | cv_qromb |   |   | 7.7 | 8.7 | 8.2 | 7.1 | 8.1 | 8
Starting | 1 | of | 10 | ionization | cycle | 7.7 | 8.8 | 8.3 | 7.2 | 8.2 | 8.1
Finished | 1 | of | 10 | ionization | cycle | 14.5 | 27 | 87.4 | 21.6 | 22.7 | 16.4
Starting | 2 | of | 10 | ionization | cycle | 14.6 | 27 | 87.4 | 21.6 | 22.8 | 16.5
Finished | 2 | of | 10 | ionization | cycle | 20.1 | 34.5 | 101.4 | 31.1 | 32.3 | 23.4
Starting | 3 | of | 10 | ionization | cycle | 20.2 | 34.5 | 101.4 | 31.1 | 32.4 | 23.5
Finished | 3 | of | 10 | ionization | cycle | 26.6 | 41.6 | 111.1 | 39.9 | 41.1 | 31
Starting | 4 | of | 10 | ionization | cycle | 26.7 | 41.7 | 111.2 | 39.9 | 41.2 | 31
Finished | 4 | of | 10 | ionization | cycle | 33.4 | 48.7 | 121 | 48.9 | 50.1 | 38.7
Starting | 5 | of | 10 | ionization | cycle | 33.4 | 48.7 | 121.1 | 48.9 | 50.2 | 38.8
Finished | 5 | of | 10 | ionization | cycle | 40.1 | 55.9 | 130.9 | 58 | 59.2 | 46.4
Starting | 6 | of | 10 | ionization | cycle | 40.2 | 55.9 | 130.9 | 58.1 | 59.3 | 46.5
Finished | 6 | of | 10 | ionization | cycle | 46.8 | 63.1 | 140.8 | 67.1 | 68.2 | 54.2
Starting | 7 | of | 10 | ionization | cycle | 46.8 | 63.1 | 140.8 | 67.1 | 68.3 | 54.2
Finished | 7 | of | 10 | ionization | cycle | 53.3 | 70.3 | 150.4 | 76.2 | 77.4 | 61.8
Starting | 8 | of | 10 | ionization | cycle | 53.3 | 70.3 | 150.4 | 76.3 | 77.4 | 61.9
Finished | 8 | of | 10 | ionization | cycle | 59.8 | 77.5 | 160.1 | 85 | 86.2 | 68.6
Starting | 9 | of | 10 | ionization | cycle | 59.9 | 77.5 | 160.1 | 85.1 | 86.2 | 68.6
Finished | 9 | of | 10 | ionization | cycle | 66.2 | 84.6 | 169.6 | 93.8 | 95 | 75.3
Starting | 10 | of | 10 | ionization | cycle | 66.2 | 84.7 | 169.7 | 93.8 | 95.1 | 75.3
Finished | 10 | of | 10 | ionization | cycle | 72.7 | 92 | 179.2 | 102.8 | 104 | 82.1
Starting | 1 | of | 10 | spectral | cycle | 72.7 | 92 | 179.2 | 102.8 | 104 | 82.1
Finished | 1 | of | 10 | spectrum | cycles | 80.5 | 99.7 | 188.8 | 112.5 | 113.7 | 90
Starting | 2 | of | 10 | spectral | cycle | 80.5 | 99.8 | 188.9 | 112.6 | 113.8 | 90.1
Finished | 2 | of | 10 | spectrum | cycles | 88.5 | 107.4 | 198.1 | 121.9 | 123.1 | 98.1
Starting | 3 | of | 10 | spectral | cycle | 88.5 | 107.4 | 198.1 | 121.9 | 123.1 | 98.1
Finished | 3 | of | 10 | spectrum | cycles | 96.3 | 114.9 | 207.2 | 131.1 | 132.3 | 105.8
Starting | 4 | of | 10 | spectral | cycle | 96.3 | 115 | 207.2 | 131.1 | 132.3 | 105.8
Finished | 4 | of | 10 | spectrum | cycles | 103.8 | 122.4 | 216.2 | 140.5 | 141.5 | 113.6
Starting | 5 | of | 10 | spectral | cycle | 103.8 | 122.5 | 216.3 | 140.6 | 141.5 | 113.6
Finished | 5 | of | 10 | spectrum | cycles | 111.5 | 130 | 225.3 | 149.8 | 150.6 | 121.3
Starting | 6 | of | 10 | spectral | cycle | 111.5 | 130.1 | 225.4 | 149.9 | 150.7 | 121.4
Finished | 6 | of | 10 | spectrum | cycles | 119.1 | 137.8 | 234.4 | 159.1 | 159.9 | 129
Starting | 7 | of | 10 | spectral | cycle | 119.2 | 137.9 | 234.5 | 159.1 | 159.9 | 129.1
Finished | 7 | of | 10 | spectrum | cycles | 126.7 | 145.3 | 243.5 | 168.2 | 169.1 | 136.5
Starting | 8 | of | 10 | spectral | cycle | 126.7 | 145.3 | 243.6 | 168.3 | 169.1 | 136.6
Finished | 8 | of | 10 | spectrum | cycles | 134.6 | 152.8 | 252.6 | 177.6 | 178.2 | 144.1
Starting | 9 | of | 10 | spectral | cycle | 134.6 | 152.8 | 252.6 | 177.6 | 178.2 | 144.2
Finished | 9 | of | 10 | spectrum | cycles | 142.8 | 160.3 | 261.6 | 186.8 | 187.4 | 151.7
Starting | 10 | of | 10 | spectral | cycle | 142.8 | 160.4 | 261.7 | 186.9 | 187.4 | 151.7
Finished | 10 | of | 10 | spectrum | cycles | 150.9 | 167.9 | 270.7 | 196.1 | 196.6 | 159.2
cv_qromb |

**Comment by Higginbottom** [comment]
Further analysis reveals that for the CV example file, of the 120s of slow down - about 30-40s is due to an increased number of levels for each ion in elem_ions file. This was done to improve the accuracy of the line cooling.
The rest is due to the slower Romberg integration in gsl - well specifically a small number of nasty integrals. This is more or less retrieved by the new qags integration scheme..

So, the question is - how important is getting line cooling right - I'd argue it is very important. 

Next question is - are we slowing down the code through the line cooling calculation, or is it due to the line opacity calculation dealing with lots more lines. In which case we may wish to have two line lists, one for opacity and one for cooling. Sounds dodgy to me tho..

**Comment by Higginbottom** [comment]
The slowdown seems to be in defining and transmitting photons.

**Comment by kslong** [comment]
I am running gprof to compare the two models.  Hope to have results a bit later

**Comment by Higginbottom** [comment]
#612 is relevant too

**Comment by kslong** [comment]
OK,  here is some information from gprof on the two runs, run on agn_short with the current dev version (only the data files are changed).  This generally confirms what Nick has said, except it looks like to me that the Romberg integration is not the issue.   (Note that this was run on a single processor on a linux machine at STScI which is somewhat slower than a modern Mac)

The top level part of the gprof file, for the old and new data.  With the new data, we are spending much more time in radiation than previously.  There are 40% more calls to radiation, nearlly all of which as shown below come from translate_in_wind.   The two routines radiation and den_config take almost 5100 s in the new run, whereas they take 915+580 or about 1500 s in the run with the old data.  

Old
```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  Ks/call  Ks/call  name    
 15.07    964.27   964.27 4833586300     0.00     0.00  fraction
 14.30   1879.70   915.43 1593179530     0.00     0.00  radiation
  9.96   2517.42   637.72 8086840819     0.00     0.00  vwind_xyz
  9.06   3097.50   580.09 2018769793     0.00     0.00  den_config
  5.16   3427.55   330.04 4986262620     0.00     0.00  coord_fraction
  4.91   3741.91   314.37 1593194800     0.00     0.00  limit_lines
  4.83   4051.12   309.21 16214850693     0.00     0.00  stuff_phot
  3.44   4271.52   220.40 24719286379     0.00     0.00  sane_check
  3.24   4479.13   207.61 8556467711     0.00     0.00  quadratic
  2.90   4664.76   185.64 1593179530     0.00     0.00  calculate_ds
  2.70   4837.80   173.04 24493216068     0.00     0.00  dot
  2.10   4971.93   134.12 4711335500     0.00     0.00  sigma_phot
  1.96   5097.53   125.60 3022585158     0.00     0.00  mean_intensity
  1.20   5174.08    76.56 1844950990     0.00     0.00  where_in_wind
  1.17   5248.88    74.80 11597258871     0.00     0.00  move_phot
  1.08   5318.10    69.22 5314589142     0.00     0.00  length
  1.00   5381.85    63.76 1593179530     0.00     0.00  kappa_ff
```

New
```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  Ks/call  Ks/call  name    
 22.30   2580.09  2580.09 2012444468     0.00     0.00  radiation
 21.74   5094.63  2514.53 4886578790     0.00     0.00  den_config
 10.58   6318.67  1224.04 10664829074     0.00     0.00  fraction
  6.86   7112.17   793.50 10126581964     0.00     0.00  vwind_xyz
  3.63   7532.51   420.34 2012459653     0.00     0.00  limit_lines
  3.51   7938.10   405.59 3012677660     0.00     0.00  coord_fraction
  3.29   8318.61   380.51 19680385004     0.00     0.00  stuff_phot
  2.92   8655.87   337.26 7513167171     0.00     0.00  sigma_phot
  2.52   8947.11   291.24 24076061202     0.00     0.00  sane_check
  2.20   9201.92   254.81 10430324520     0.00     0.00  quadratic
  2.04   9438.11   236.19 2012444468     0.00     0.00  calculate_ds
  1.95   9663.54   225.43 30104780452     0.00     0.00  dot
  1.28   9811.12   147.58 3547144994     0.00     0.00  mean_intensity
  0.98   9924.53   113.41 14132103477     0.00     0.00  move_phot
  0.80  10016.92    92.39 2297818450     0.00     0.00  where_in_wind
  0.75  10103.51    86.59 6634547932     0.00     0.00  length
  0.71  10185.07    81.56 2012444468     0.00     0.00  translate_in_wind
  0.68  10263.62    78.56 2012444468     0.00     0.00  kappa_ff
```


The section of the detailed report for the routine radiation

Old
```
              915.43 1648.75 1593179530/1593179530     translate_in_wind [6]
[9]     40.1  915.43 1648.75 1593179530         radiation [9]
              580.07    0.00 2018701585/2018769793     den_config [17]
              251.27  225.19 3186359060/8086840819     vwind_xyz [12]
              114.72   51.36 4029972243/4711335500     sigma_phot [27]
               51.00   91.26 1593179530/1593179530     kappa_ind_comp [30]
               63.76    0.00 1593179530/1593179530     kappa_ff [41]
               33.85   10.85 1593179530/1593179530     kappa_comp [43]
               43.27    0.00 4853603168/24719286379     sane_check [23]
               25.06   13.91 1704276397/2300421295     renorm [42]
               31.09    0.00 1630211819/16214850693     stuff_phot [21]
               22.51    3.50 3186359060/24493216068     dot [26]
               10.51    0.00 1630211819/11597258871     move_phot [39]
                4.51    3.98 37032289/37032289     update_banded_estimators [75]
                7.87    0.00 1704276397/2173744737     stuff_v [69]
                0.87    5.17 111096867/461970312     project_from_xyz_cyl [56]
                2.70    0.00 1593179530/1597729480     check_plasma [96]
                0.46    0.00 37032289/1630211819     klein_nishina [58]
```

New
```
             2580.09 3856.17 2012444468/2012444468     translate_in_wind [4]
[9]     55.6 2580.09 3856.17 2012444468         radiation [9]
             2514.45    0.00 4886414444/4886578790     den_config [12]
              315.38  259.36 4024888936/10126581964     vwind_xyz [13]
              176.85   34.19 3939719821/7513167171     sigma_phot [22]
               65.03  120.55 2012444468/2012444468     kappa_ind_comp [34]
               78.56    0.00 2012444468/2012444468     kappa_ff [43]
               73.95    0.00 6113540182/24076061202     sane_check [26]
               42.04   12.48 2012444468/2012444468     kappa_comp [48]
               31.91   15.96 2126754635/2851511009     renorm [45]
               39.65    0.00 2050547857/19680385004     stuff_phot [23]
               30.14    0.06 4024888936/30104780452     dot [32]
               16.46    0.00 2050547857/14132103477     move_phot [39]
               10.45    0.00 2126754635/2733963191     stuff_v [76]
                5.14    4.08 38103389/38103389     update_banded_estimators [80]
                0.89    5.08 114310167/574551345     project_from_xyz_cyl [57]
                3.07    0.00 2012444468/2019008119     check_plasma [98]
                0.46    0.00 38103389/2050547857     klein_nishina [60]
-----------------------------------------------
```
den_config is almost entirely called by radiation.  den_config does not have any daughters, so the time in den_config really is spent there.

**Comment by Higginbottom** [comment]
Well, thats very odd - since i can see on wall time that changing from romberg to qags shaves significant time off the runs..
I have made the change to the integrator in pull #624 which is just running travis checks.

**Comment by kslong** [comment]
Regarding how to go forward, I'd suggest making Standard78 and Standard80 part of the distribution ... though we may not want to call them this.  I don't want to give up the speed, unless necessary.

But we do need to continue to work to try to speed the program up and this should tell us wehre.

**Comment by kslong** [comment]
@Higginbottom  I have rerun the regression tests on the unloaded laptop I had run them on earlier after having pulled the latest version of dev, with your integration fixes in place.  It reduced the run time of agn_short from 2100 to 1700 seconds, but not to the 1100 s we used to get.  There is more info on the Evernote page about this.  Some of the other regression examples did get fairly close to what we used to have, so there is something about agn_short that is more affected by the switch to standard80 than the others, maybe it is simply that there are more interactions.  

I also reran the profiling test above on standard80 on a linux machine.  Interestingly there, there was almost no change in the overall run time (possibly because I am not able to control the loading of the linux server).

**Comment by Higginbottom** [comment]
I have made a new standard file standard80_reduced which uses the old elem_ions_73 file with only 5 levels per ion.
I have also modified the examples file to use this. They run much faster, the only change is to the agn_short as one would expect (since it runs much more slowly - presumably all the lines are doing something.


![agn_short](https://user-images.githubusercontent.com/3329213/67947466-9bb12c00-fbdb-11e9-8744-d931347d055a.png)
